### ISSUE TITLE
tt-macro doesn't work

### ISSUE THREAD

**Issue by jbclements** [rust-lang/rust#5846]
I haven't finished investigating this, so I'm making an issue just so I don't forget it:

```rust
macro_rules! foo (
    ($x:tt) => {type t = ($x) int;})

foo!(~)
```

**Comment by kmcallister** [comment]
The modernized example

```rust
macro_rules! foo {
    ($x:tt) => (type t = $x<int>;)
}

foo!(Box);
```

gives me

```
/tmp/foo.rs:2:26: 2:28 error: expected type, found `an interpolated tt`
/tmp/foo.rs:2     ($x:tt) => (type t = $x<int>;)
                                       ^~
```

which looks like #22819.

**Comment by BigBoneDaddy** [comment]
@kmcallister Is this still problem?

**Comment by huonw** [comment]
https://github.com/rust-lang/rust/issues/16036 notes that another hacky work-around is a nested macro invocation, which avoids needing to define/export dummy macros (although one will want to be careful about giving that internal macro a fairly unique name to avoid trampling on any other macros in scope, unfortunately :disappointed: ).

```rust
macro_rules! foo {
    ($x:tt) => (
        macro_rules! internal {
            () => { type t = $x<i8>; }
        }
        internal!();
    )
}

foo!(Box);
```

**Comment by steveklabnik** [comment]
Updated example:

```rust
macro_rules! foo {
    ($x:tt) => (type t = $x<i32>;)
}

foo!(Box);
```

error:

```
error: expected type, found `Box`
 --> <anon>:2:26
2 |>     ($x:tt) => (type t = $x<i32>;)
  |>                          ^^
```

**Comment by brson** [comment]
Triage: This is an ancient P-low bug but per @sanxiyn it actually comes up in the wild a lot - macro writers have to know how to do this workaround. Can this be escalated? Is there a solution somebody could mentor?

**Comment by nrc** [comment]
Discussed at lang meeting, we believe this should remain p-low since we don't know a very easy fix. It should be fixed by converting macro_rules to expand to pure tokens then parsing, rather than mixing the two steps (which we hop would be back-compat).

cc @jseyfried

**Comment by jseyfried** [comment]
Fixed in #34908.
